File: dart/lib/src/twitter/get_twitter_registry_key.dart
import 'package:solana/solana.dart' hide RpcClient;
import '../constants/addresses.dart';
import '../utils/get_domain_key_sync.dart';

/// Get Twitter registry key for a Twitter handle
///
/// This function mirrors js/src/twitter/getTwitterRegistryKey.ts
///
/// [twitterHandle] - The Twitter handle (without @)
///
/// Returns the public key for the Twitter registry
Future<Ed25519HDPublicKey> getTwitterRegistryKey(String twitterHandle) async {
  final hashedTwitterHandle = getHashedNameSync(twitterHandle);

  final keyString = await getNameAccountKeySync(
    hashedTwitterHandle,
    nameParent: twitterRootParentRegistryAddress,
  );

  return Ed25519HDPublicKey.fromBase58(keyString);
}

File: js/src/twitter/getTwitterRegistryKey.ts
import { PublicKey } from "@solana/web3.js";
import { TWITTER_ROOT_PARENT_REGISTRY_KEY } from "../constants";
import { getHashedNameSync } from "../utils/getHashedNameSync";
import { getNameAccountKeySync } from "../utils/getNameAccountKeySync";

/**
 * Derives the user-facing name registry key for a Twitter handle.
 *
 * @param twitter_handle Twitter handle without the `@` prefix
 * @returns Derived Twitter name-registry account public key
 *
 * @example
 * ```ts
 * const key = await getTwitterRegistryKey("bonfida");
 * ```
 */
export async function getTwitterRegistryKey(
  twitter_handle: string,
): Promise<PublicKey> {
  const hashedTwitterHandle = getHashedNameSync(twitter_handle);
  return getNameAccountKeySync(
    hashedTwitterHandle,
    undefined,
    TWITTER_ROOT_PARENT_REGISTRY_KEY,
  );
}
